### ISSUE TITLE
Urls with # as an action are not matched

### ISSUE THREAD

**Issue by chemouna** [airbnb/DeepLinkDispatch#37]
I have a case with this url : "http://www.example.com/#/{query}" but its not matched, i looked at the code and it's because encodedPath()  in DeepLinkUri that looks for queries after ? or # . 
    ```int pathEnd = delimiterOffset(url, pathStart, url.length(), "?#");```

Here's a failing test case for it : 
```
  @Test public void urlWithHash() {
    DeepLinkEntry entry = deepLinkEntry("http://example.com/{any}/{nb}");  
                                         //or deepLinkEntry("http://example.com/#/{nb}");
    Map<String, String> parameters = entry.getParameters("http://example.com/#/2");
    assertThat(parameters.get("nb")).isEqualTo("2");
  }
```
is there a way to have  an escape character or something for such a case ? 

Thank you.

**Comment by felipecsl** [comment]
According to the [URL spec](https://url.spec.whatwg.org/#non_relative-path-state), `#` delimits the start of the fragment, so it is not part of the path. Right now we only support interpolating parameters in the URL host and path.

**Comment by skptl** [comment]
@DeepLink("scheme://{object_type}/{id}") was working with version 1.3.0, but with v1.4.0 it throws a run-time error as follows:

java.lang.RuntimeException: Unable to start activity ComponentInfo{com.package.example/com.airbnb.deeplinkdispatch.DeepLinkActivity}: java.util.regex.PatternSyntaxException: Error in {min,max} interval near index n:

**Comment by felipecsl** [comment]
@skptl can you provide a sample Uri that triggers this exception?

**Comment by skptl** [comment]
Yes, sure!

URL: staging://letters/561e88f79aab50f91f0005e1

@DeepLink("staging://{objectType}/{id}")
public class ObjectActivity{
...
}

java.lang.RuntimeException: Unable to start activity ComponentInfo{com.staging/com.airbnb.deeplinkdispatch.DeepLinkActivity}: java.util.regex.PatternSyntaxException: Error in {min,max} interval near index 11:

staging://{objecttype}/([a-zA-Z0-9_#!+%-~,.$]*)$

**Comment by skptl** [comment]
Thanks @felipecsl, I appreciate your quick response!

**Comment by MichaelEvans** [comment]
@skptl I think the fix for your problem is in #54

**Comment by felipecsl** [comment]
Fixed by #54
